### ISSUE TITLE
[BUG] watch: rebuild always rebuilds all (or dependent) containers

### ISSUE THREAD

**Issue by nocive** [docker/compose#10863]
### Description

I've tried using the watch feature with a standard webpack setup using sync & rebuild, similar to what the documentation shows.

I have a `x-develop` section like this in my container definition:

```
x-develop:
  watch:
    - action: sync
      path: <path>/webpack-out
      target: /app
      ignore:
        - node_modules/
    - action: rebuild
      path: <path>/package.json
```

While `sync` works as expected, for `rebuild` action things behave oddly or at least different from what my expectation would be.

Every time `package.json` is changed, docker compose is rebuilding all my project containers (or perhaps just the dependent ones?).

With a big project with >50 containers such as ours, this makes `rebuild` completely unusable as every file change will trigger a rebuild of multiple containers (this can easily add up to many minutes of build time).

```
+ exec docker compose alpha watch my-app-container
watch command is EXPERIMENTAL
watching [<path>/webpack-out <path>/package.json]
change detected on <path>/package.json
change detected on <path>/package.json
change detected on <path>/package.json
change detected on <path>/package.json
change detected on <path>/package.json
change detected on <path>/package.json
change detected on <path>/package.json
change detected on <path>/package.json
change detected on <path>/package.json
change detected on <path>/package.json
change detected on <path>/package.json
change detected on <path>/package.json
change detected on <path>/package.json
change detected on <path>/package.json
change detected on <path>/package.json
change detected on <path>/package.json
change detected on <path>/package.json
Rebuilding my-app-container after changes were detected:
  - <path>/package.json
 
( ...proceeds to rebuild many containers, not just my-app-container... )
```

### Steps To Reproduce

1. Use a project with multiple containers and dependencies and a rebuild section for some-app-container
2. Build and start some-app-container
3. Run docker compose alpha watch some-app-container
4. Change a file targeted in the rebuild section of some-app-container
5. Observe docker compose rebuilding multiple containers, not just some-app-container

### Compose Version

```Text
Docker Compose version v2.20.2
```

### Docker Environment

```Text
Client:
 Version:    24.0.2
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  0.11.0
    Path:     /usr/lib/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.20.2
    Path:     /home/jsaraiva/.docker/cli-plugins/docker-compose

Server:
 Containers: 798
  Running: 39
  Paused: 0
  Stopped: 759
 Images: 367
 Server Version: 24.0.2
 Storage Driver: overlay2
  Backing Filesystem: btrfs
  Supports d_type: true
  Using metacopy: true
  Native Overlay Diff: false
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 0cae528dd6cb557f7201036e9f43420650207b58.m
 runc version: 
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: xxx
 Operating System: xxx
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 31.01GiB
 Name: jps-x2
 ID: xxx
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username: xxx
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by milas** [comment]
This will be fixed in the next version of Compose by #10956.

Thanks for the report!

**Comment by nocive** [comment]
works like a charm in latest 2.23.0 version, thanks :+1:

**Comment by alkami-jdalrymple** [comment]
Is this on the roadmap for `Docker Desktop`?

**Comment by alkami-jdalrymple** [comment]
Still experiencing this in 2.22.0:
```
$ docker compose version
Docker Compose version v2.22.0-desktop.2
```

With the following:
```yaml
services:
  container_one:
    image: foo:bar
    command: "test"
    develop:
      watch:
        - action: sync
          path: .
          target: /app
  container_two:
    image: foo:bar
    command: "test"
    build:
      context: .
      target: bar
      tags:
        - "foo:bar"
    develop:
      watch:
        - action: rebuild
          path: .
```

Unfortunately, this restarts both `container_one` and `container_two`, instead of just `container_one`.

**Comment by Swivelgames** [comment]
Still experiencing this in 2.22.0:
```
$ docker compose version
Docker Compose version v2.22.0-desktop.2
```

With the following:
```yaml
services:
  container_one:
    image: foo:bar
    command: "test"
    develop:
      watch:
        - action: sync
          path: .
          target: /app
  container_two:
    image: foo:bar
    command: "test"
    build:
      context: .
      target: bar
      tags:
        - "foo:bar"
    develop:
      watch:
        - action: rebuild
          path: .
```

Unfortunately, this restarts both `container_one` and `container_two`, instead of just `container_one`.

**Comment by nocive** [comment]
@Swivelgames it's not fixed in 2.22.0.

**Comment by damiandzwigaj** [comment]
still problem in v2.23.0
